### ISSUE TITLE
The Future of Babel-sublime

### ISSUE THREAD

**Issue by Thom1729** [babel/babel-sublime#355]
Paging:
- @zertosh
- @xtuc

# Introduction

I'm a software developer who works largely in web application development. I write a lot of JavaScript, much of it for React. I've also been a Sublime user for years and more recently a contributor to the core packages and the author of several of my own. In particular, I am currently the primary active contributor to the core JavaScript syntax. I also functionally wrote most of the new Vue syntax, though I'm not a Vue user myself.

I have been a long-time Babel-sublime user, but a few months ago, frustrated with Babel-sublime's many bugs, I switched to my own JavaScript package, [JS Custom](https://github.com/Thom1729/Sublime-JS-Custom).

What I want, as a developer who uses React, is a syntax that:

- Impeccably highlights standard JavaScript.
- Also highlights JSX and a few other nonstandard features.
- Complies with Sublime's [scope naming guidelines](https://www.sublimetext.com/docs/3/scope_naming.html) and other accepted best practices.
- Is compatible to the greatest possible degree with first- and third-party packages and tools designed for the core JavaScript syntax.

# Background

Babel-sublime is by far the most popular package for JSX development in Sublime Text. According to [Package Control](https://packagecontrol.io/packages/Babel), it has been downloaded 382,589 times. Countless websites and articles recommend Babel-sublime.

Babel originated in a time when Sublime's core JavaScript syntax mouldered from neglect. [Benvie's JavaScriptNext package](https://github.com/Benvie/JavaScriptNext.tmLanguage) provided better highlighting that included new ECMAScript features that many developers used. The Babel syntax built upon that, continuing to add standards-track ECMAScript features while providing support for JSX and Flow. Eventually, it supplanted JavaScriptNext as the premier JavaScript package for Sublime.

# Current Status

It's been more than two years since Babel-sublime received a major update. Meanwhile, the core JavaScript package has been almost completely overhauled. First, the old `tmLanguage` was replaced by a YAML-based syntax derived (like Babel-sublime) from JavaScriptNext. But since the `sublime-syntax` conversion, virtually every part of that syntax has been modified or rewritten, mush of it within the past year. Nearly all known bugs have been fixed, and its support for new ECMAScript features is top-notch.

Babel-sublime suffers from a large list of known bugs, including many that would be nearly impossible to fix in a `tmLanguage`. Although JavaScriptNext was once the best syntax out there, it suffered from serious structural deficiencies that made it nearly impossible to correctly distinguish regexps from division operators, arrays from property accesses, and so on. Fixing these problems in Sublime's core syntax required [a massive effort](https://github.com/sublimehq/Packages/pull/1009) spanning many months in order to implement a brand-new architecture designed especially for the `sublime-syntax` format. Fixing the same issues in Babel-sublime would require a similar effort.

However, there is no chance of the core syntax replacing Babel-sublime because the core syntax is deliberately conservative when it comes to experimental and nonstandard features. It is possible that it might someday implement JSX, but not Flow, Styled Components, or many of the other nonstandard features that Babel provides and upon which countless users rely.

# The Problems with Babel-sublime

## Structure

The Babel-sublime syntax is fundamentally to a `tmLanguage`. Although its editable source is YAML, it is compiled to a `tmLanguage` and is subject to all of the limitations of that format.

Converting the syntax to a `sublime-syntax` file would be trivial. The problem is that correct JavaScript highlighting requires a radically different structure than the one that the `tmLanguage` format enforces. The necessary structure is heavily stack-based; the core syntax requires over three hundred `push`, `pop`, and `set` rules. On the other hand, Babel-sublime relies largely on regular expression capture groups: it uses 223 to the core syntax's 58. Most of these would have to be rewritten in a totally different form.

## Bugs

As a result of these structural deficiencies, Babel inevitably has trouble with:

- Automatic semicolon insertion. (#354, #330, #309)
- Regular expression identification and parsing. (#349, #342)
- Arrow functions. (#343, #340)

In addition, many common constructs will be highlighted incorrectly if valid whitespace or comments are encountered. These problems can only be fixed by a complete restructuring.

## Performance

In addition, the old `tmLanguage`-based structure is bad for performance. Babel takes more than twice as long to parse the [jQuery source](https://code.jquery.com/jquery-3.3.1.js) as does the core JavaScript syntax. Experience has proved that manipulating contexts is fast, while regular expression captures are slow.

## Conformance

Babel-sublime and the core syntax have diverged substantially over the years. Babel-sublime does not conform to the scope naming guidelines and this results in visibly different highlighting for plain JavaScript. This defeats the purpose of having standard highlighting.

## Tests

Babel-sublime has no syntax tests. This makes it very difficult to edit, because it's impossible to know whether a seemingly innocuous change will break working code. The syntax tests for Sublime's core JavaScript syntax contain over four thousand assertions refined over the course of years. Without these tests, last year's overhaul of the core syntax would have been virtually impossible.

# Moving Forward

Fixing these issues manually in the existing syntax is technically possible, but probably impractical. First, someone would have to write literally thousands of syntax tests to document Babel-sublime's existing behavior. Only then could the mammoth task of overhauling the syntax begin. In the end, it is overwhelmingly likely that the result would effectively be a brand-new syntax.

Instead, I think that what we need to do is effectively "rebase" from the core syntax. There's no need to duplicate all of the work of highlighting vanilla JavaScript; the core syntax does that just fine, is actively maintained, and has thousands of tests already. The added features that Babel-sublime provides, like JSX and Flow, can be added on top of this. The result is best-of-class JavaScript highlighting plus the critical extra features that Babel-sublime users rely upon.

I am confident in this approach because I've already done it. [JS Custom](https://github.com/Thom1729/Sublime-JS-Custom) extends the core JavaScript syntax to add support for JSX, Flow, and many other features. It's a true extension -- the new features exist in patches that are applied to an unaltered copy of the core syntax and compiled. This means that it will never diverge from the core syntax, and integrating upstream improvements is trivially easy. It also means that you can turn individual features like JSX or Flow on and off, a feature that will allow us to implement TypeScript support alongside Flow.

There are two ways that I can see this working. One is that JS Custom could subsume Babel-sublime. This would be the simplest approach. The other is that Babel-sublime could ship a compiled, Babel-customized syntax generated with JS Custom. This would be the most "lightweight" in that Babel-sublime wouldn't include the full apparatus of JS Custom, but this would limit Babel-sublime's potential for expansion (into TypeScript, etc).

In either case, work would have to be done to get both packages ready. I'm still working on the test framework for JS Custom, which runs Sublime's core syntax tests alongside many new ones. The most important thing is ensuring a graceful transition for existing Babel-sublime users. JS Custom is already nearly a complete drop-in replacement for Babel-sublime, but the last mile is the most important.

# Final Thoughts

I hope that I don't seem presumptuous recommending such radical changes. I want to see Babel-sublime succeed rather than languish, and I'm willing, able, and eager to do the work needed to get it there.

I would greatly appreciate input from Babel-sublime users and contributors; I don't want anyone to feel left in the dust.

Questions, comments, concerns, complaints?

**Comment by Thom1729** [comment]
I presume that @xtuc would have said something by now if he had any major objections, so I'll continue.

First, @zertosh, I want to express my appreciation for all the work you've done for this syntax over the years. I relied heavily on it myself for a long time, as do my coworkers and countless JavaScript developers. That this syntax would eventually diverge from core was inevitable; only in the past year has the technology been developed that might keep it in sync. And I can commiserate with the difficulty of dealing with JavaScript syntax; it's a challenge even with the full power of `sublime-syntax`, and some problems simply can't be solved because the syntax is *nondeterministic* context-free.

Overhauling the core JavaScript syntax was like repairing a boat at sea; every change had the potential to break something critically important. For the core syntax, there were compelling reasons to do that anyway; with Babel, we should be able to jump straight to the end.

That said, every improvement is a change, and this would be a very big change. So the first item on the roadmap should be understanding the scope of the change. Specifically, we should identify:

- Features that Babel provides that the core syntax does not.
- Places where Babel differs from the core syntax for the same functionality, and whether that difference ought to be kept or eliminated.
- Third-party packages that are compatible with Babel-sublime.

Then, we can evaluate JS Custom's ability to produce a syntax that satisfies those requirements. Compatibility with Sublime-babel has been a top priority, but I'm sure that there are some ways in which it falls short. These must be remedied on the JS Custom side (and, in some cases, in the core syntax).

Once this is done, we can create a new major version of Babel-sublime with a new `sublime-syntax` compiled using JS Custom. This would inevitably be a significant change. Before the stable release, we'd need to document:

- Known major bug fixes, especially those that might be [mistaken for new bugs](https://github.com/sublimehq/Packages/issues/1309).
- Other places where we change behavior to match the core syntax.
- Any extant bugs, including those due to parser limitations.

Once the new version has stabilized, we could either integrate the full functionality of JS Custom into Babel-sublime, or (probably more likely) reframe JS Custom as a customizable version of Babel.

At this point, Babel-sublime would be set up to immediately benefit from improvements to the core JavaScript syntax. As new ECMAScript features are added, they could be added to Babel-sublime (via JS Custom) even before the core syntax adopted them.

Does this sound like a reasonable plan of action?

**Comment by xtuc** [comment]
I couldn't react before sorry. I agree and I'd like to thank you for your efforts. 

I'll let @zertosh decide. Let me know if I can help with something.

**Comment by Thom1729** [comment]
I've opened #356 for the first phase of work.

**Comment by Thom1729** [comment]
Updates:

The core JavaScript syntax has open PRs for [decorators](https://github.com/sublimehq/Packages/pull/1483) and [template string keybinds](https://github.com/sublimehq/Packages/pull/1490). When these are accepted, the core syntax will support every standard JavaScript feature that this package does. I'm optimistic that a new stable version would be out soon; we would use this as the base for any upgrades to Babel-sublime.

I'm finishing up the next major version of [JS Custom](https://github.com/Thom1729/Sublime-JS-Custom). There are a number of syntax improvements, especially for flow typing, but the marquee feature is a complete set of syntax tests. (I've been doing some work on the [UnitTesting](https://github.com/SublimeText/UnitTesting/) package to support this, and I'm hoping that [sublime_lib](https://github.com/SublimeText/sublime_lib) will also be out in time for the JS Custom release.)

**Comment by getogrand** [comment]
@zertosh How about inviting @Thom1729 to babel team for maintaining `babel-sublime`?
`babel-sublime` is already defacto standard package in Sublime Text ecosystem and it's not easy to make people move to the another.

In late 2008, `merb` was merged into `rails`. 
I hope `JS Custom` will be merged into `babel-sublime` like that.

**Comment by Thom1729** [comment]
Updates:

- [Sublime Text 3.1](https://www.sublimetext.com/3) is out, with a number of JavaScript improvements.
- Although ES6 Decorators did not make the 3.1 release, they have been [merged](https://github.com/sublimehq/Packages/pull/1483) into the `master` branch and will be present in future releases.
- [JS Custom 1.0.13](https://github.com/Thom1729/Sublime-JS-Custom) is out, with support for more ES6 features and a number of other improvements and fixes.
- [sublime_lib](https://github.com/SublimeText/sublime_lib/) is nearing a stable release. JS Custom 1.1 will rely on several sublime_lib features.

Possible roadmap:

- Publish a beta release of sublime_lib (this week).
- Refactor JS Custom 1.1 to use sublime_lib.
- Finish JS Custom syntax test system.
- Create a baseline "Babel" configuration for JS Custom.
- Publish the compiled syntax in a new Babel-sublime branch.
- Test out the new syntax for a month or two.
- Release the new syntax to everyone.

**Comment by Thom1729** [comment]
After an excessive delay:

- [sublime_lib](https://github.com/SublimeText/sublime_lib) is out.
- [yaml_macros_engine](https://github.com/Thom1729/yaml-macros-engine) version 2.0.0 is out, featuring a substantially rewritten core.
- [JS Custom](https://github.com/Thom1729/Sublime-JS-Custom) version 2.0.0 is out, relying on the above dependencies, with major refactoring and dynamically generated syntax tests.
- A number of improvements have been made to the core JavaScript syntax, including overhauled support scopes.

There are therefore no roadblocks to implementing the changes suggested in this thread. The next steps should be:

- Create a `next` branch.
- Choose JS Custom configuration options for the Babel configuration.
- Build that configuration and deploy it to the branch.
- Create a prerelease tag.

@zertosh: If you create a branch, I can PR it. Alternatively, I can do it myself if you add me to the repo.

**Comment by Thom1729** [comment]
@zertosh @xtuc ping.

**Comment by KevinSjoberg** [comment]
First of all, big thanks to you @Thom1729. I truly appreciate all the hard work you've put in into making JavaScript a seamless experience in Sublime Text. ❤️ 

Did we ever get the ball rolling on this one? Is the plan still to incorporate changes from [JS Custom](https://github.com/Thom1729/Sublime-JS-Custom)? Is there anything we, as the community, can do to help?

**Comment by Thom1729** [comment]
I'm still interested in moving forward on this. The most conservative approach would be for one of the maintainers to create a `next` branch, and I then can send in a PR. Or, if someone adds me to the repository, I can do that myself.

**Comment by Thom1729** [comment]
@zertosh @xtuc

Are you still interested in maintaining the package? If not, then I'd be willing to fork it and transfer ownership in Package Control.

**Comment by KevinSjoberg** [comment]
Thank you, @Thom1729 for pushing the needle forward! ❤️

**Comment by KevinSjoberg** [comment]
Ping @zertosh @xtuc

**Comment by nicolo-ribaudo** [comment]
@Thom1729 Sorry for the super long delay; I wasn't watching this repo. I'll talk about this with the team and report back in the next few days

**Comment by Thom1729** [comment]
The Babel team has added me to the repository, so I'm closing this issue in favor of #379.

**Comment by KevinSjoberg** [comment]
@Thom1729 great news! 🎉
